Handle plurals for clozes

wclausen opened the pull request:
This PR attempts to apply a very basic solution to the problem of plurals not being clozed out (and plural forms of boring words still being clozed out).

The approach is to simply remove any trailing 's' characters, which is about the most basic form of lemmatization for nouns.

One thing to note is that the lemmatization needs to be applied to every keyword (including the named entities) because it is used in the cloze logic for comparing a given word in the sentence against the lemmatized keyword. We can change the logic if this is overly complicated so that the cloze logic clozes a word by considering both the old way and new way. Or we can say this is too much hassle and bag it.

The pr also updates the boring words list with stuff that I have seen in recent reviews

z1lc commented on tests/test_highlight_clozer.py:
```diff
@@ -53,7 +59,7 @@ def test_clean_keyword():
         "ugly duckling",
         "Duchess of Cambridge",
-        "LeBron James",
+        "LeBron Jame",  # lemmatization will strip the trailing 's'
```
Seems undesirable -- are we relying on `_clean_keyword` in the Person pipeline? Also does this mean that in the below test (`test_cloze_out_multiword_keyword`) the keyword we'll be clozing out would be LeBron Jame?

wclausen replied:
The code for extracting persons is independent of the code for clozing highlights, so this should be fine.

Re: what keyword is being clozed out. The keyword being clozed out would be "Lebron James". What the cloze contains is always the full world. The decision about whether to cloze the full word is based on checking the cleaned/lemmatized word. Think of the lemma information as being a mapping from a given word in the highlight back to a canonical form of that word. When we loop through the words in the highlight, instead of checking that the actual word exactly matches, we check if a given word has the same canonical form as the keyword (which is now always in canonical form).

wclausen replied:
ofc i agree that this example is undesirable. using the nlp's ability to do lemmatization would avoid this, but seemed like more effort to set up. And this doesn't seem to affect correctness for the vast majority of situations.

It's not perfect obv, but it is simple (for what we are trying to do) and done.

z1lc replied:
Ah, great explanation, makes sense.

z1lc reviewed (approved):
Not sure I fully understand how these methods impact the overall pipeline but it seems like the end to end tests still are good so probably fine. Maybe we need to add some more end 2 end tests in the future

wclausen commented:
I addressed some of the thinking for this change in response to comments in the code. For clarity/future understanding, I think it's worth being more explicit about what is changing and why.

Example to help clarify:
sentence: `The mountains are very tall. This mountain is the tallest`

Before this change:
keyword: `mountains`
Result: `The {{c1::mountains}} are very tall. This mountain is the tallest`
Why is `mountain` not also clozed? Because the decision to cloze is based on exact matching with full word. Keyword was `mountains` and `mountain` is a different word.

After this change
keyword: `mountain` (keywords will be in lemmatized form, aka won't have punctuation or trailing `s` chars)
Result: `The {{c1::mountains}} are very tall. This {{c1::mountain}} is the tallest`
The reason `mountain` and `mountains` is clozed out is that when we look at `mountains` we now map it to the lemmatized form when deciding whether to cloze it out. So the cloze code sees `mountains` as being `mountain` which is equal to the keyword so the whole word is clozed out.

**Limitations**
I have implemented an extremely basic and weak form of lemmatization. It's literally just trimming punctuation and removing trailing `s` characters. Removing the `s` character can potentially result in some words being mapped to the same lemma by mistake (I can't think of any examples, but it is a real theoretical concern). It will occasionally map plurals to words that aren't the singular form (e.g. "stories" is not mapped to "story" but instead to "storie", like I mentioned this is very weak lemmatization). It also will sometimes change non-plural words (e.g. "LeBron James" is mapped to "LeBron Jame"). The reason we are ok with these issues is that they are extremely uncommon and this approach will catch a lot of the plurals correctly (e.g. it solves the issue where "mountains" is clozed but "mountain" is not) and the mapping will only extremely rarely create a result that is worse than what we had before (e.g. even in the "LeBron James" case, the resulting cloze will contain the whole word and produce the same card we had before). It is a real 80/20, which necessarily means it doesn't handle everything perfectly.
